pkg/ako-operator/akodeploymentconfig_cluster_mapping_helper.go

```diff
@@ -0,0 +1,149 @@
+	ctx context.Context,
+	kclient client.Client,
+	log logr.Logger,
+	obj *akoov1alpha1.AKODeploymentConfig) (*clusterv1.ClusterList, error) {
+	// get all clusters can be selected by this akodeploymentconfig's cluster selector
+	selector, err := metav1.LabelSelectorAsSelector(&obj.Spec.ClusterSelector)
+	if err != nil {
+		return nil, err
+	}
+	var clusters clusterv1.ClusterList
+	if err := kclient.List(ctx, &clusters, []client.ListOption{
+		client.MatchingLabelsSelector{Selector: selector},
+	}...); err != nil {
+		return nil, err
+	}
+	// remove clusters that:
+	// 1. not ready
+	// 2. management cluster
+	// 3. previously selected by other adc objects
+	var newItems []clusterv1.Cluster
+	var allErrs []error
+	for _, cluster := range clusters.Items {
+		if !SkipCluster(&cluster) {
+			// management cluster can't be selected by other adc objects
+			// except the management cluster AKODeploymentConfig
+			if cluster.Namespace == akoov1alpha1.TKGSystemNamespace &&
+				obj.Name != akoov1alpha1.ManagementClusterAkoDeploymentConfig {
+				continue
+			}
+			adcName, exist := cluster.Labels[akoov1alpha1.AviClusterLabel]
+			// if cluster is already selected by other customized adc objects, skip
+			// only clusters selected by default adc with empty selector object can be overrided
+			if exist && adcName != obj.Name {
+				if !isDefaultADC(adcName) || !defaultADCHasEmptySelector(ctx, kclient) {
+					continue
+				}
+			}
+			newItems = append(newItems, cluster)
+		}
+	}
+	clusters.Items = newItems
+	return &clusters, kerrors.NewAggregate(allErrs)
+}
+
+// GetAKODeploymentConfigForCluster return the akodeloymentconfig object which selects
+// current cluster
+func GetAKODeploymentConfigForCluster(
+	ctx context.Context,
+	kclient client.Client,
+	log logr.Logger,
+	cluster *clusterv1.Cluster) (*akoov1alpha1.AKODeploymentConfig, error) {
+	// list all the akodeploymentconfig objects
+	var akoDeploymentConfigs akoov1alpha1.AKODeploymentConfigList
+	if err := kclient.List(ctx, &akoDeploymentConfigs, []client.ListOption{}...); err != nil {
+		log.Error(err, "Failed to list all AKODeploymentConfig objects")
+		return nil, err
+	}
+	// find which adc matches current cluster
+	var defaultAdc akoov1alpha1.AKODeploymentConfig
+	for _, akoDeploymentConfig := range akoDeploymentConfigs.Items {
+		if selector, err := metav1.LabelSelectorAsSelector(&akoDeploymentConfig.Spec.ClusterSelector); err != nil {
+			log.Error(err, "Failed to convert label sector to selector")
+		} else if selector.Empty() {
+			if isDefaultADC(akoDeploymentConfig.Name) {
+				defaultAdc = akoDeploymentConfig
+			}
+		} else if selector.Matches(labels.Set(cluster.GetLabels())) {
+			log.Info("cluster is selected by akodeploymentconfig", "adc", akoDeploymentConfig.Name)
+			return &akoDeploymentConfig, nil
+		}
+	}
+	// only default adc with empty selector can select all clusters and return
+	if defaultAdc.Name == akoov1alpha1.WorkloadClusterAkoDeploymentConfig {
```

Comment:
What if cluster is mgmt-cluster?

Reply:
mgmt cluster will be selected by `install-ako-for-mgmt` since only mgmt cluster has the label: `cluster-role.tkg.tanzu.vmware.com/management: ""` which can match `install-ako-for-mgmt`'s cluster selector